### ISSUE TITLE
mix_compile_unless_stale_config wrongly says that the configs files changed

### ISSUE THREAD

**Issue by jn-jairo** [phoenixframework/phoenix#5956]
### Environment

* Elixir version (elixir -v): Elixir 1.17.3 (compiled with Erlang/OTP 27)
* Phoenix version (mix deps): phoenix 1.7.14 
* Operating system:
  * docker: hexpm/elixir:1.17.3-erlang-27.1.1-debian-bookworm-20240926-slim
  * host: Ubuntu 20.04.6 LTS 5.4.0-198-generic x86_64 GNU/Linux

### Actual behavior

For some reason when the reload code runs, in some cases the file `_build/dev/lib/<project>/.mix/compile.elixir` doesn't exist and the `mix_compile_unless_stale_config()` shows the message "You must restart your server after changing the following files", because the `Mix.Utils.extract_stale()` compares the last_modified numbers and because the file doesn't exist it is 0 and all configs last_modified numbers are greater than 0, thus making it "stale", but the configs files didn't change.

### Expected behavior

Don't force to restart the server when the configs didn't change.

### My workaround

Modifying the `mix_compile_unless_stale_config()` to check if the manifests files exist and compile if they don't exist, solves the problem.

I don't know why it happens, it may be some racing condition, I'm using phoenix_live_view 1.0.0-rc.7 I believe it has some reloading, I don't know if it is related.

And it is an umbrella project.

To test if it works I'm modifying the page tile on the live view, like:

```elixir
defmodule HelloWeb.Blog.PostLive.Index do
  use HelloWeb, :live_view

  alias Hello.Blog
  alias Hello.Blog.Post

  ...

  defp apply_action(socket, :index, _params) do
    socket
    |> assign(:page_title, "Posts") # Changing the page title triggers the error
    |> assign(:post, nil)
  end

  ...

end
```

I believe this is not the only change that triggers the error, because I have to restart the server a lot due to this error, but it's an easy one to test.

Regardless of what is making the `compile.elixir` don't exist at some times, I believe the code should account for it and just compile the project.

**Comment by jn-jairo** [comment]
I tried another thing, just filtering the existing files also solves the problem:

```elixir
  defp mix_compile_unless_stale_config(compilers, compile_args, timestamp, path) do
    manifests = Mix.Tasks.Compile.Elixir.manifests() |> Enum.filter(&File.exists?/1)
```

It's simpler than what I was doing before, just one line changed.

**Comment by josevalim** [comment]
Thanks for the report. Do you have a way to reproduce the issue? If your Phoenix app is running, then it is not possible for `_build/dev/lib/<project>/.mix/compile.elixir` to be missing, unless it was manually removed, and I'd say starting the app is indeed the right way to go.

**Comment by jn-jairo** [comment]
Alright, I found the problem, it is a racing condition, phoenix is not the one removing the `compile.elixir`, the sequence of events is:

1. Start phoenix server, it compiles and the `compile.elixir` exists.
2. Edit file in IDE, the IDE linter runs `mix credo list --format=oneline --strict` and it compiles at the same time phoenix server is trying to compile, what removes and recreates the `compile.elixir`, so for a brief moment the `compile.elixir` doesn't exist, and sometimes the `mix_compile_unless_stale_config` runs while the `compile.elixir` doesn't exist.

I guess when the compilation takes longer the racing condition happens more often.

It happens when editing on the IDE because both the linter (credo) and phoenix recompile triggers when the file changes.

I don't know how to solve it, because the file changed, so credo will recompile when it runs and the phoenix server will also recompile.

Racing conditions are hard to solve 🙃

**Comment by josevalim** [comment]
Thanks for debugging! This is fixed in Elixir v1.18 as we now have cross-OS lock, so we can close this. :)

**Comment by jn-jairo** [comment]
Just for notice, I checked the new cross-os lock, but it is not useful in this case because the key for the lock is different, the key is the build_path but the path is different inside the docker container and on the host and the locks are on the tmp folder which are different inside/outside docker.

But don't worry, that problem is too specific, I can work without that credo linter and if I really wanted it, I could hack the plugin to run the credo command inside the docker.

I'm just writing it here for the others with the same problem that find this page.

I don't expect any solution for this problem from the Elixir team, a customizable key won't work because of the lock folder and even with the key and folder problem solved the tcp port would be a problem in docker, because the port in docker should be open so the host and container can communicate, in this case the docker container and the host are basically different machines, it is a dead end and not worth to try to solve it, in my opinion.
